source/common/event/scaled_range_timer_manager_impl.cc

```diff
@@ -255,12 +259,14 @@ void ScaledRangeTimerManagerImpl::onQueueTimerFired(Queue& queue) {
 
   // Pop and trigger timers until the one at the front isn't supposed to have expired yet (given the
   // current scale factor).
+  queue.processing_timers_ = true;
   while (!timers.empty() &&
          computeTriggerTime(timers.front(), queue.duration_, scale_factor_) <= now) {
     auto item = std::move(queue.range_timers_.front());
     queue.range_timers_.pop_front();
     item.timer_.trigger();
   }
+  queue.processing_timers_ = false;
```

Comment:
So processing_timers_ needs to be per queue since otherwise the queue not empty invariant would not hold if a timer from queue A removes a timer from queue B.  It would be good to also cover that case in tests.

Also, is there code somewhere that verifies the invariant that all active queues are non-empty?

Reply:
> Also, is there code somewhere that verifies the invariant that all active queues are non-empty?

I don't believe there is currently. The number of queues and sizes is not currently exposed. I'm okay to keep this PR focused on the bugfix and add additional testing in a follow-on.

Reply:
023d9ec adds a test for cancelling a timer from a different queue in the context of a timer callback.

Reply:
cc: @asraa 

@akonradi Please do followup.  Regarding ASSERTs, I think failing this one will result in a crash in the next line.

https://github.com/envoyproxy/envoy/blob/bf84a7b51bbe8a9496cc5c4bb0df14606f9545f0/source/common/event/scaled_range_timer_manager_impl.cc#L244